Tasks/FtpUpload/ftpuploadtask.ts

```diff
@@ -0,0 +1,278 @@
+        var remaining: number = files.length - total;
+        message = message + getFinalStatusMessage() + '\nunprocessed files & directories: '+ remaining;
+    }
+    tl.setResult(tl.TaskResult.Failed, message);
+}
+
+function uploadFiles() {
+    tl.debug('files to process: ' + files.length);
+
+    createRemoteDirectory(remotePath); // ensure root remote location exists
+
+    files.forEach((file) => {
+        tl.debug('file: ' + file);
+        var remoteFile: string = preservePaths ?
+            path.join(remotePath, file.substring(rootFolder.length)) :
+            path.join(remotePath, path.basename(file));
+
+        remoteFile = remoteFile.replace(/\\/gi, "/"); // use forward slashes always
+        tl.debug('remoteFile: ' + remoteFile);
+
+        var stats = tl.stats(file);
+        if (stats.isDirectory()) { // create directories if necessary
+            createRemoteDirectory(remoteFile);
+        } else if (stats.isFile()) { // upload files
+            if (overwrite) {
+                uploadFile(file, remoteFile);
+            } else {
+                remoteExists(remoteFile, function(exists: boolean){
+                    if(!exists){
+                        uploadFile(file, remoteFile);
+                    } else {
+                        filesSkipped++;
+                        checkDone('skipping file: ' + file + ' remote: ' + remoteFile + ' because it already exists');
+                    }
+                });
+            }
+        }
+    });
+}
+
+// only called in no overwrite mode
+function remoteExists(remoteFile : string, callback){
+    var remoteDirname = path.normalize(path.dirname(remoteFile));
+    var remoteBasename = path.basename(remoteFile);
+    //todo -- optimize to cache information
+    c.list(remoteDirname, function (err, list) {
+        if(err){
+            if (err.code == 550){ // standard not found error
+                callback(false);
+            } else { // some other error -- just return false --
+                callback(false);
+            }
+            return false;
+        } else {
+            for (var remote of list) {
+                if (remote.name == remoteBasename) {
+                    callback(true);
+                    return;
+                }
+            }
+            callback(false);
+            return true;
+        }
+    });
+}
+
+function createRemoteDirectory(remoteDirectory: string) {
+    tl.debug('creating remote directory: ' + remoteDirectory);
+    c.mkdir(remoteDirectory, true, function (err) {
+        if (err) {
+            c.end();
+            failTask('Unable to create remote directory: ' + remoteDirectory + ' due to error: ' + err);
+        }
+        directoriesProcessed++;
+        checkDone('remote directory successfully created/verified: ' + remoteDirectory);
+    });
+}
+
+function uploadFile(file: string, remoteFile) {
+    tl.debug('uploading file: ' + file + ' remote: ' + remoteFile);
```

Comment:
The FTP connection seems to handle all of this gracefully such that clients don't need to worry about it.